xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -497,5 +443,175 @@ protected Helper delegate() {
         return orcaReportingHelperWrapper.asHelper();
       }
     }
+
+    private final class PriorityManager {
+
+      private final List<List<XdsLocality>> priorityTable = new ArrayList<>();
+      private Map<XdsLocality, LocalityInfo> localityInfoMap = ImmutableMap.of();
+      private int currentPriority = -1;
+      private ScheduledHandle failOverTimer;
+
+      /**
+       * Updates the priority ordering of localities with the given collection of localities.
+       * Recomputes the current ready localities to be used.
+       */
+      void updateLocalities(Map<XdsLocality, LocalityInfo> localityInfoMap) {
+        this.localityInfoMap = ImmutableMap.copyOf(localityInfoMap);
+        priorityTable.clear();
+        for (XdsLocality newLocality : localityInfoMap.keySet()) {
+          int priority = localityInfoMap.get(newLocality).priority;
+          while (priorityTable.size() <= priority) {
+            priorityTable.add(new ArrayList<XdsLocality>());
+          }
+          priorityTable.get(priority).add(newLocality);
+        }
+
+        if (currentPriority >= priorityTable.size()) {
```

Comment:
This is verbose and misleading. You always want to run `updatePriorityState` to find the new smallest priority upon localities are refreshed. All the following lines can be reduced to 

```java
currentPriority = priorityTable.size() - 1;
for (int p = 0; i < priorityTable.size(); p++) {
  updatePriorityState(p);
}
```

Reply:
I'm not disagreeing the logic can be simplified.
But `currentPriority` is changed in two separate methods, in your proposed change, in this method, `currentPriority` is not at a meaningful state. 
I want to clearly see `currentPriority` be in a meaningful state in every single method. Otherwise it's misleading.

Reply:
```java
if (currentPriority >= priorityTable.size()) {
  currentPriority = priorityTable.size() - 1;
}
```
is not meaningful, at least for me.

Set `currentPriority` to `priorityTable.size() - 1` and then search for the available locality with smallest priority. Why is this not meaningful? The idea is the same as setting the upper bound to `Integer.MAX_VALUE` and then do a binary search.

Reply:
Why is the 'if then' not meaningful for you? If the priorotyTable size becomes smaller - even smaller than the current priority, then you obviously have to decrease the priority value. It's obvious even you don't know about the algorithm detail.

Setting an upper bound is not obvious to me, you have to know the algorithm detail, and have to know that after a loop of the separate method `updatePriorityState()` it will change `currentPriority` to a value that makes sense, and that the loop of  `updatePriorityState()` depends an upper bound. If `updatePriorityState()`, `gotoPriority()` and `failOver()` were all inlined in this method, I could barely accept this, because then I could see clearly how the field `currentPriority` is changed in one block.

Reply:
Anyway, setting `currentPriority` here is temporary, its eventual value is set after iterating over the new `priorityTable` to find the available locality with smallest priority. 

Setting `currentPriority` to `priorityTable.size() - 1` only when it exceeds the table size (use the old value if not). You have two different cases for setting a temporary value, this is not intuitive.

I am ok if you want to keep this.

Reply:
Actually the upper bound logic is wrong. If changed to the upper bound logic, it will do the wrong thing at least for the initial update:

```
event detail - expected state
EDS update: P0 sz1; P1 sz2; P2 sz3; P3 sz4 - P0 C, P1 N/A, P2 N/A, P3 N/A
```

It will change `currentPriority` to 3, and the updated state will be `P0 C, P1 C, P2 C, P3 C`
```
io.grpc.xds.LocalityStoreTest > multipriority FAILED
    unexpected (3): sz2, sz4, sz3
    ---
    expected      : [sz1]
    but was       : [sz2, sz1, sz4, sz3]
        at io.grpc.xds.LocalityStoreTest.multipriority(LocalityStoreTest.java:912)
```

The algorithm is NOT simply "reset an upper bound first, and then search for the available locality with smallest priority".

Reply:
Try
```java
currentPriority = -1;
failOver();
```

Reply:
It breaks a lot of exiting single priority tests, not only `multipriority()` test. And I quite don't follow this logic.

Reply:
I tried with 

```java
void updateLocalities(Map<XdsLocality, LocalityInfo> localityInfoMap) {
        this.localityInfoMap = ImmutableMap.copyOf(localityInfoMap);
        priorityTable.clear();
        for (XdsLocality newLocality : localityInfoMap.keySet()) {
          int priority = localityInfoMap.get(newLocality).priority;
          while (priorityTable.size() <= priority) {
            priorityTable.add(new ArrayList<XdsLocality>());
          }
          priorityTable.get(priority).add(newLocality);
        }

        cancelFailOverTimer();
        currentPriority = -1;
        failOver();
}
```

I don't see any tests break.

All you need upon locality update is to find the new `currentPriority`. Isn't this the goal of the algorithm?

Reply:
No. It shouldn't `cancelFailOverTimer()` regardless. In some case the timer should remain.

Reply:
I should add a test case like

9 sec passes, EDS update without priority change, then after 1 sec timer should expire.

Thanks for catching a corner case.

Reply:
What's the problem for always researching the new available locality with smallest priority? That's the spec as we discussed previously.

Reply:
I think handling this corner case is flexible, we could just recompute `currentPriority` every time instead of handle the case when updated localities are exactly the same as previous ones.

Reply:
> What's the problem for always researching the new available locality with smallest priority?

There's no problem with that. But if there is no READY locality in the new update, we use the existing `currentPriority`; and if further that the existing `currentPriority` >=  `priorityTable.size()`, which does not make send, we should decrease it to `priorityTable.size() - 1`.

Reply:
Upon receiving new locality update, `priorityTable` has been updated with new localities, `currentPriority` is no long valid as it points to a new set of localities. It needs to be recomputed anyway.

Reply:
Let me think if we can ignore the corner case, because that is a really rare case and does not matter much. And then the code will be simplified a lot as you suggested.

Reply:
So suppose we apply the simple code path: In the corner case, i.e. in case the existing current priority is still CONNECTING, and before the 10 sec timer expires, if a new EDS update comes without priority change (e.g. only endpoint change), then it will cancel the current 10 sec timer and eagerly try the next priority immediately. Seems not big of a deal.

Reply:
> it points to a new set of localities. It needs to be recomputed anyway.

PS: This is also a minor edge case.

Reply:
> So suppose we apply the simple code path: In the corner case, i.e. in case the existing current priority is still CONNECTING, and before the 10 sec timer expires, if a new EDS update comes without priority change (e.g. only endpoint change), then it will cancel the current 10 sec timer and eagerly try the next priority immediately. Seems not big of a deal.

Why not just recomputes `currentPriority`? This can be generalized to all cases.

Reply:
In any way it affects only corner case, which we can ignore. I changed the code to reset `currentPriority` and recompute `currentPriority`.